torch/_inductor/sizevars.py

```diff
@@ -243,46 +227,85 @@ def guard_equals(self, left: Expr, right: Expr) -> Expr:
+        """
         if left == right:
             return True
-        if self.size_hint(left - right) == 0:
-            self.guard_equals(left, right)
-            return True
-        return False
+        expr = sympy.Eq(left, right)
+        simplified = self.shape_env._maybe_evaluate_static(expr)
+
+        return is_expr_static_and_true(simplified)
 
-    # See Note [maybe_guard vs guard]
-    def maybe_guard_list_equals(self, left: List[Expr], right: List[Expr]) -> bool:
-        """if left==right, guard on that fact and return true"""
+    # See Note - [On Statically Known]
+    def statically_known_list_equals(self, left: List[Expr], right: List[Expr]) -> bool:
+        """
+        Returns a bool indicating if it is sound to optimize as if left and right lists are equal.
+        """
         if len(left) != len(right):
             return False
-        if all(self.size_hint(a - b) == 0 for a, b in zip(left, right)):
-            for a, b in zip(left, right):
-                self.guard_equals(a, b)
+        if all(self.statically_known_equals(l, r) for l, r in zip(left, right)):
             return True
         return False
 
-    # See Note [maybe_guard vs guard]
-    def maybe_guard_leq(self, left: Expr, right: Expr) -> bool:
-        try:
-            if self.size_hint(left) > self.size_hint(right):
-                return False
-        except TypeError:
-            return False
-        self.guard_leq(left, right)
-        return True
+    # See Note - [On Statically Known]
+    def statically_known_leq(self, left: Expr, right: Expr) -> bool:
+        """
+        Returns a bool indicating if it is sound to optimize as if left is less than or equal to right.
+        """
+        expr = left <= right
+        simplified = self.shape_env._maybe_evaluate_static(expr)
 
-    # See Note [maybe_guard vs guard]
-    def maybe_guard_lt(self, left: Expr, right: Expr) -> bool:
-        try:
-            if self.size_hint(left) >= self.size_hint(right):
-                return False
-        except TypeError:
-            return False
-        self.guard_lt(left, right)
-        return True
+        return is_expr_static_and_true(simplified)
+
+    # See Note - [On Statically Known]
+    def statically_known_lt(self, left: Expr, right: Expr) -> bool:
+        """
+        Returns a bool indicating if it is sound to optimize as if left is less than right.
+        """
+        expr = left < right
+        simplified = self.shape_env._maybe_evaluate_static(expr)
+
+        return is_expr_static_and_true(simplified)
+
+    # See Note - [On Statically Known]
+    def statically_known_multiple_of(self, numerator: Expr, denominator: Expr) -> bool:
+        """
+        Return a bool indicating if it is sound to optimize for the numerator being a multiple of the denominator.
+        """
+        if sympy.gcd(numerator, denominator) == denominator:
+            # can prove it symbolically
+            return True
```

Comment:
hmm, I hope we don't actually need this; `_maybe_evaluate_static` hopefully ends up doing something equivalent to gcd